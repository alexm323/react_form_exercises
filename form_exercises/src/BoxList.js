import React, { useState } from 'react';
import { v4 as uuid } from 'uuid';
import Box from './Box'
import NewBoxForm from './NewBoxForm'

const BoxList = () => {
    // we need to keep track of the state of the list so we can add boxes to it 
    const [boxes, setBoxes] = useState([])
    // we want to make a function that we can pass down to our Box component as a prop using the data that we get back from our NewBoxForm component 
    const addBox = (newBox) => {
        // update the array but we are not mutating the original , we are copying it and adding a new property/box to it
        setBoxes(boxes => [...boxes, { ...newBox, id: uuid() }])
    }
    // add a method that we can use to remove a box with their specific id
    const removeBox = (id) => {
        setBoxes(boxes => boxes.filter(box => box.id !== id));

    }
    return (
        <div>
            <h2>Box Factory</h2>
            <NewBoxForm addBox={addBox} />
            <div>
                {boxes.map(box => < Box handleRemove={removeBox} backgroundColor={box.backgroundColor} id={box.id} height={`${box.height}px`} width={`${box.width}px`} />
                )}
            </div>


        </div>

    )
}
export default BoxList;