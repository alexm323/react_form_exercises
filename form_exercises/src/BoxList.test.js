import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

// smokebox and snapshot
it("renders without crashing", () => {
    render(<BoxList />);
})

it("matches snapshots", () => {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

// create a function that allows us to add a box to the boxlist component
function addBox(boxList, height = "2", width = "2", color = "peachpuff") {
    // capture the inputs as the user sees them by their labels 
    const colorInput = boxList.getByLabelText("Color");
    const heightInput = boxList.getByLabelText("Height");
    const widthInput = boxList.getByLabelText("Width");

    // capture the submit button by its text 
    const button = boxList.getByText("Add Color Box")

    // no we can add data (with the use of fireEvent) to the input fields to add a box with that data 
    fireEvent.change(colorInput, { target: { value: color } });

    fireEvent.change(widthInput, { target: { value: height } });

    fireEvent.change(heightInput, { target: { value: width } });
    // submit
    fireEvent.click(button);
}

it("can add a new box", () => {
    // before we add anything there should be an empty boxlist component 
    const boxList = render(<BoxList />);

    expect(boxList.queryByText("Remove Box")).not.toBeInTheDocument();

    addBox(boxList, 'red', '100', '100')

    expect(boxList.queryByText("Remove Box")).toBeInTheDocument();
})