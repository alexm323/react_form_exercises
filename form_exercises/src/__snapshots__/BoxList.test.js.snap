// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Box Factory
    </h2>
    <form>
      <label
        forhtml="color"
      >
        Color
      </label>
      <input
        aria-label="Color"
        id="color"
        name="color"
        type="text"
        value=""
      />
      <label
        forhtml="height"
      >
        Height
      </label>
      <input
        aria-label="Height"
        id="height"
        name="height"
        type="text"
        value=""
      />
      <label
        forhtml="width"
      >
        Width
      </label>
      <input
        aria-label="Width"
        id="width"
        name="width"
        type="text"
        value=""
      />
      <button>
        Add Color Box
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
